zcash_client_backend/src/data_api/chain.rs

```diff
@@ -220,6 +223,198 @@ pub trait BlockSource {
+///    block_cache.insert(compact_blocks.clone()).unwrap();
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 2);
+///
+///    // Find highest block in the block cache
+///    let get_tip_height = block_cache.get_tip_height(None).unwrap();
+///    assert_eq!(get_tip_height, Some(BlockHeight::from_u32(2)));
+///
+///    // Read from the block cache
+///    let blocks_from_cache = block_cache.read(&range).unwrap();
+///    assert_eq!(blocks_from_cache, compact_blocks);
+///
+///    // Truncate the block cache
+///    block_cache.truncate(BlockHeight::from_u32(1)).unwrap();
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 1);
+///    assert_eq!(
+///        block_cache.get_tip_height(None).unwrap(),
+///        Some(BlockHeight::from_u32(1))
+///    );
+///
+///    // Delete blocks from the block cache
+///    let rt = tokio::runtime::Runtime::new().unwrap();
+///    rt.block_on(async {
+///        block_cache.delete(&range).await.unwrap();
+///    });
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 0);
+///    assert_eq!(block_cache.get_tip_height(None).unwrap(), None);
+/// ```
+#[cfg(feature = "sync")]
+pub trait BlockCache: BlockSource + Send + Sync {
+    /// Retrieves contiguous compact blocks specified by the given `range` from the block cache.
+    ///
+    /// Returns `Ok(Vec<CompactBlock>)` on success, otherwise returns an error.
+    ///
+    /// Short reads are allowed, meaning that returning fewer blocks than requested should not
+    /// return an error as long as all blocks are sequentially continuous in height.
+    ///
+    /// # Errors
+    ///
+    /// This method should return an error if there are gaps in the requested range of blocks,
+    /// indicating there are blocks missing from the cache.
+    fn read(&self, range: &ScanRange) -> Result<Vec<CompactBlock>, Self::Error>;
+
+    /// Finds the height of the highest block known to the block cache within a specified range.
+    /// If `range` is `None`, returns the tip of the entire cache.
+    ///
+    /// Returns `Ok(Some(BlockHeight))` on success, otherwise returns an error.
+    /// If no blocks are found in the cache, returns Ok(`None`).
+    fn get_tip_height(&self, range: Option<&ScanRange>)
+        -> Result<Option<BlockHeight>, Self::Error>;
+
+    /// Inserts a vec of compact blocks into the block cache.
+    ///
+    /// Returns `Ok(())` on success, otherwise returns an error.
+    fn insert(&self, compact_blocks: Vec<CompactBlock>) -> Result<(), Self::Error>;
+
+    /// Removes all cached blocks above a specified block height.
+    ///
+    /// Returns `Ok(())` on success, otherwise returns an error.
+    fn truncate(&self, block_height: BlockHeight) -> Result<(), Self::Error>;
+
+    /// Deletes a range of compact blocks from the block cache.
+    ///
+    /// Returns a `JoinHandle` from a `tokio::spawn` task.
+    fn delete(&self, range: &ScanRange) -> JoinHandle<Result<(), Self::Error>>;
```

Comment:
`tokio::task::JoinHandle` requires the `tokio/rt` feature flag, which seems excessive for the library (requiring us to at a library level depend on a specific runtime). I think all we actually need here is a `Box<dyn Future>` we can then await? `JoinHandle` implements `Future`, so a concrete implementation of this trait should be able to satisfy it with `tokio`.

We might also want to use the `async-trait` crate here to enable effectively returning `impl Future` instead, by making this:
```suggestion
    async fn delete(&self, range: &ScanRange) -> Result<(), Self::Error>;
```

Reply:
Ok yes I prefer this to be async and was how I implemented originally. I will put this back but use the `aync-trait` crate so its compatible with rust v1.65